<a href="https://colab.research.google.com/github/vvtrip/ml_manifestations/blob/master/stats/4_stats_with_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import numpy as np
from scipy import stats
import statistics

# Basics

**Descriptive Statistics** 
 - it summaries a dataset, which helps in gaining insights, and making inferences about a dataset.
 - In general, we compute statistical measures of one or more samples, related to a population, and draw conclusions about the population.
 - Descriptive Statistics involves estimating centrality measures and measures of dispersion.

**Centrality Measures**
 - It determine the center of a dataset.
 - The three major centrality measures are: mean, median, and mode.

**Mean**
 - it is the sum of all values divided by a total number of values, of a data set.
 - mean function of numpy is used to compute mean of an array of numbers

**Median**
 - it is the value that separates the given data set into two halves.
 - median function of numpy library can be used to compute median of a data set.

**Mode**
 - it is the value that appears most often in a dataset is called the mode value.
 - mode function of scipy.stats module can be used for computing mode of a given data set.

**Measures of Dispersion**
 - it provides insights on the spread of given dataset.
 - Major measures of dispersion are: range, percentile, inter-quartile range, standard deviation, variance, skewness, and kurtosis.


**Range**
 - it is the difference between maximum and minimum values of the dataset.
- import numpy as np
- s1 = np.array([86, 47, 45, 47, 40])
- print(np.ptp(s1))

**Percentile**
 - it refers to a value, below which lies given the percentage of data points.

 - E.g., 45th percentile refers to a value below which 45% of data points are found.
 - percentile function of numpy can be used to compute a single or multiple percentiles.

**Quartiles**
 - Three Quartiles namely, Q1 Q2 and Q3, split the entire dataset into four equal parts.
 - Each part contains 25% of data.

**Inter Quartile Range (IQR)**
 - it refers to difference between third quartile (Q3) and first quartile (Q1).
 - iqr method from scipy.stats can be used for calculating it.

**Variance**
 - it is defined as the average of squared differences, of each data point from dataset's mean.

**Standard Deviation**
 - it is square root of variance.
 - var and std functions of numpy can be used for computing variance and standard deviation respectively.
 - By default, the functions assume that the dataset represents entire population.
-  To represent a sample, derived from a population, ddof parameter(delta degree of freedom is set to 1.

**Skewness**
- it determines whether the majority of data points are present on one side of the distribution.
 - A positive value represents right skewed distribution; a negative value represents left skewed one, zero represent unskewed distribution.

**Kurtosis**
 - Kurtosis indicates how much of data is concentrated around mean or shape of the probability distribution.
 - It can be estimated using kurtosis function of scipy.stats module.
 - By default, it uses Fisher’s definition. This can be changed to Pearson by setting fisher parameter to False.

In [2]:
s = [26, 15, 8, 44, 26, 13, 38, 24, 17, 29]
print(f"Mean is {np.mean(s)}")
print(f"Meadian is {np.median(s)}")
print(f"Mode is {stats.mode(s)}, {stats.mode(s)[0][0]}")
print(f"without interpolation 1st and 3rd quartiles are {np.percentile(s,[25,75])}")
print(f"without interpolation IQR through scipy.stats is {stats.iqr(s, rng=(25,75))}")
print(f"with interpolation 1st and 3rd quartiles are {np.percentile(s,[25,75], interpolation='lower')}")
print(f"with interpolation IQR through scipy.stats is {stats.iqr(s, rng=(25,75), interpolation='lower')}")
print(f"Skewness is {round(stats.skew(s),2)}")
print(f"Kurtosis is {round(stats.kurtosis(s),2)}")

Mean is 24.0
Meadian is 25.0
Mode is ModeResult(mode=array([26]), count=array([2])), 26
without interpolation 1st and 3rd quartiles are [15.5  28.25]
without interpolation IQR through scipy.stats is 12.75
with interpolation 1st and 3rd quartiles are [15 26]
with interpolation IQR through scipy.stats is 11
Skewness is 0.36
Kurtosis is -0.77


# Random Variables, Random Numbers and Random Distributions

A random number is a number, chosen by chance from a distribution.
Python provides a lot of modules, which deal with random numbers.
- random module of python standard library.
- random module of Numpy
- stats module of Scipy

random module of numpy has utilities, which generate arrays of random numbers.
- E.g.: rand function generates uniformly distributed numbers from range [0, 1]

- rand function with no arguments generate a single random value.

- By passing arguments, it generates a random array of specified size.

In [3]:
import numpy as np
print(np.random.rand())
# generates a 2*3 array
print(np.random.rand(2,3))

0.8947447978672337
[[0.0256188  0.46978763 0.71970412]
 [0.34350112 0.63733897 0.46798065]]


- In statistics, you select items randomly from a population, either with or without a replacement.
- This can be achieved with choice method

In [4]:
import numpy as np
print(np.random.choice([11, 22, 33], 2, replace=False))

[33 11]


**Random Seeding**
- Seed is an important concept when it comes to reproducibility. If you are working with random numbers and you would want to peers to validate your results, i.e., they should also get the same random sequence as you did, you can set the seed to a particular value and send the seed value to your peers.
- seed is a number that sets the initial state of random number generator.
- Setting a seed, helps in generating the same sequence of random numbers, repeatedly.
- seed method of a random module can be used to set a seed as shown in below example.

In [5]:
import numpy as np

np.random.seed(100)
print(np.random.rand())

np.random.seed(100)
print(np.random.rand())

0.5434049417909654
0.5434049417909654


**Random Variables**
 - In probability theory, the set of all possible outcomes of a random experiment is known as sample space.
 - Probabilities of all outcomes of the experiment define the probability distribution.
 - A random variable is a variable that takes real numbers or integers and map each value to one of the outcomes of sample space. 
 - E.g.: In an experiment of tossing a coin, the sample space is {'Head', 'Tail'} and a possible random variable takes the value 0 for head and 1 for the tail.

**Probability Distributions**
 - There are two types of probability distributions namely discrete and continuous that take integer and real values, respectively.
 - scipy.stats module provides classes that represent random variables, corresponding to a large number of probability distributions.
 - E.g: the class norm represent normal continuous random variable, and binom represent binomial discrete random variable.

- scipy.stats module provide a lot of methods for created discrete and continuous random variables.
- Commonly used methods are :
- pdf / pmf : Probability distribution function (continuous) or probability mass function (discrete).
- cdf : Cumulative distribution function.
- sf : Survival function (1 – cdf).
- rvs : Creating random samples from a distribution.
- The following example defines a normal continuous random variable of mean 1.0 and std 2.5.
- It also estimates probabilities and cumulative probabilities at -1, 0 and 1.
- The example also generates six random numbers from defined normal distribution.

In [6]:
from scipy import stats

x = stats.norm(loc=1.0, scale=2.5)

print(x.pdf([-1, 0, 1]))

print(x.cdf([-1, 0, 1]))

print(x.rvs((2,3)))

[0.11587662 0.14730806 0.15957691]
[0.2118554  0.34457826 0.5       ]
[[-0.40409689 -3.12269376  1.88668613]
 [-0.96516083  0.42031951  1.5199392 ]]


In [7]:
# create a normal distribution with mean 32 and standard deviation 4.5
# set the random seed to 1, and create a random sample of 100 elements from the above distribution
# compute aboslute difference between the sample mean and distribution mean

import numpy as np
from scipy import stats
np.random.seed(1)

x = stats.norm.rvs(loc=32, scale=4.5, size=100)
print(round(abs(np.mean(x)-32),2))

0.27


In [8]:
# Simulate a random experiment of tossing a coin 10000 times and determine the count of Heads.
# Hint: Define a binomial distribution with n = 1 and p = 0.5

import numpy as np
from scipy import stats
np.random.seed(1)

x = stats.binom(1, 0.5)
y = x.rvs(10000)
print(np.bincount(y)[0])

4990


# Hypothesis testing with scipy


- Hypothesis Testing is a methodology for evaluating if a claim is acceptable or not, based on data.
- In a Hypothesis Testing, a Null Hypothesis (Ho) represents currently accepted the state of knowledge, and an Alternative Hypothesis (Ha) represents a new claim which challenges the currently accepted state of knowledge.
- The null hypothesis and the alternative hypothesis are mutually exclusive.

**Steps Involved in Hypothesis Testing**
- The following steps are involved in a Hypothesis Testing:
 - Define the null hypothesis and the alternative hypothesis.
 - Select a test statistics whose probability distribution function can be found under the null hypothesis.
 - Collect data.
 - Compute the test statistics from the data and calculate its p-value under the null hypothesis.
 - Null hypothesis is rejected if the p-value is lower than predetermined significance value.

**Choosing Test Statistics**
 - In hypothesis testing, selecting a test statistics is the most difficult part.
 - The methods used for performing t-test are shown below.
 - stats.ttest_1samp: Tests if the mean of a population is a given value.
 - stats.ttest_ind: Tests if the means of two independent samples are equal.
 - stats.ttest_rel: Tests if the means of two paired samples are equal.

**Example 1**
 - Let's consider a common hypothesis: Mean of a population is equal to certain value.
 - In reality, we estimate mean and variance of a sample and calculate the test statistic.
 - If population variance is identified, then it is reasonable to consider that test statistic is normally distributed.
 - If population variance is unknown, sample variance is used, and test statistic follows t distribution.

In [9]:
# Consider a normal population with mean 0.8 and standard deviation 0.5.
# Define the null hypothesis as Mean of the population is 1.0.
# Let's calculate t-statistic and p-value

from scipy import stats
import numpy as np
np.random.seed(1)

mu, sigma = 0.8, 0.5
X = stats.norm(mu, sigma)

# Deriving a sample
n = 100
X_sample = X.rvs(n)

# Computing test statistic
t, p = stats.ttest_1samp(X_sample, 1.0)
print(round(t,2), p)

-3.82 0.00023686273495632666


In previous example, the obtained t-statistic value is 3.82 and p-value is 0.0000023686....

Since p-value is very low and less than the significance level 0.05, you can reject the null hypothesis and infer that the mean of the population is not 1.

In [10]:
# Let's consider another problem, where the null hypothesis states 
# that the population means of two random variables are equal.

# The below example derives two samples from different populations 
# and verifies the claim that their population means are equal.

X1 = stats.norm(0.25, 1.0)
X2 = stats.norm(0.50, 1.0)
np.random.seed(1)

X1_sample = X1.rvs(100)
X2_sample = X2.rvs(100)

t, p = stats.ttest_ind(X1_sample, X2_sample)
print(round(t,2), p)

-2.65 0.008722880782836376


In previous example, the obtained t-statistic value is -2.65 and p-value is 0.008722...

Since p-value is very low and less than the significance level 0.05, you can reject the null hypothesis and state that population means of both samples differ.

In [11]:
import numpy as np
from scipy import stats

# The following are samples represent life satisfaction score(through a methodology)
# of older adults and younger adults respectively. Compute t-statistics amnd display the
# t-score and p-value 
# sol - these are independent samples

s1 = [45, 38, 52, 48, 25, 39, 51, 46, 55, 46]
s2 = [34, 22, 15, 27, 37, 41, 24, 19, 26, 36]

t, p = stats.ttest_ind(s1, s2)

print(t, p)

4.257546665558161 0.0004736633119019225


In [12]:
#Write your code here

import numpy as np
from scipy import stats

# following are samples of chocolate chip consumption by 10 rats with and 
#  without electrical simulation. calculate t-statistics and provide t-score and p-value
# sol - these are related samples

s1 = [12, 7, 3, 11, 8, 5, 14, 7, 9, 10]
s2 = [8, 7, 4, 14, 6, 7, 12, 5, 5, 8]

t, p = stats.ttest_rel(s1, s2)

print(t, p)

1.315587028960544 0.2208380130273219


# Statistical Modelling

 - A Statistical Model is a mathematical equation, which explains the relationship between dependent variables (Y) and independent variables (X).
 - In general, a model is written as Y = f(X)
However, in reality, an element of uncertainty is expected due to factors such as measurement noise. Hence the aforementioned equation can be rewritten as Y = f(X) + e   'e' is residual error

**Statistical Modelling**
- Statistical Modelling deals with creating models that attempt to explain the data best.
- The simplest model is a linear model represented as Y = B0 + B1*X + e, where the coefficients, B0 and B1 are the parameters of the model and e is normally distributed residual error.
- A linear regression model assumes that residuals are independent and normally distributed.
- The model is fitted to data using ordinary least squares approach.

**Linear Models**
 - In most of the linear regression models, a dependent variable y is written as :
 - a linear combination of the response variables X i.e y = B0 + B1*X1 + ... + Bn*Xn, or
functions of the response variables i.e y = B0 + B1*X + B2*x^2 + ... + Bn*X^n, or
 - models that have a linear component i.e y = B0 + B1*sin(X1) + B2*cos(X2)

**Non-Linear Models**
 - Other than linear regression models, statistical modeling can be used to build non-linear models.

 - Errors of dependent variable follow a distribution other than normal distribution.

 - Examples of non-linear models are: Binomial Regression and Poisson Regression.

 - In most of the cases, the non-linear models are generalized to linear models.

**Design Matrices**
 - In reality, you choose a model and fit the available data to it.

 - Once a model is chosen, design matrices y and X are constructed, and the regression problem is written, in matrix form, as y = XB + e.

 - where y is the vector of dependent variables, X is the vector of independent variables, B is a vector of coefficients, and e is the residual (error).

 - Thus obtained design matrices are passed as inputs to the chosen model.

**Statistical Modelling with StatsModels**
 - statsmodels library supports several types of statistical models.

 - All of them follow a similar usage pattern.

 - A Statistical model is represented by a model class.

 - A model can be initiated with,

  - given design matrices of dependent and independent variables, or with
  - given patsy formula and data frame or dictionary-like object.

**Step 1: Creating a Model**
 - An instance of a model class is created in either of the following ways.

 - model = sm.MODEL(y, X) or
 - model = smf.model(patsy_formula, data)
 
 where MODEL and model refer to model names such as OLS, GLM, ols, glm, etc.

 Uppercase names take design matrices as arguments, and lowercase names take Patsy formulas and data frames as arguments.

**Step 2: Fitting a Model**

In order to fit the model with data, fit method is invoked on the created model, as shown below.

 - result = model.fit()

The fit method returns a result object, which has methods and attributes for further analysis.


**Step 3: Viewing Model Summary**
 - The summary method of result object produces a summary text that describes the result of the fit, as shown below.

 - print(result.summary())

 - The displayed summary text varies for each statistical model and provides information of various statistical parameters.

**Step 4: Analyzing the Model Further**
Other than viewing summary statistics, you can perform activities like,

 - Determining fitted values,
 - Predicting the dependent variable values for new independent variable values.
 - Checking if residuals of fitted models follow a normal distribution or not.

# Statistical modelling with Patsy

**Constructing Design Matrices**

The below example shows calculation of design matrices, y, and X, for the considered linear model Y = B0 + B1*X1 + B2*X2 + B3*X1*X2.

In [13]:
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
X = np.vstack([np.ones(5), x1, x2, x1*x2]).T

print(y)
print(X)

# Thus obtained design matrices (y and X) can be passed to regression methods for obtaining the coefficient vector.

[1 2 3 4 5]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


**Design Matrices with patsy**
 - patsy, a  python library, allows defining a model in simpler easily.

 - It also constructs relevant design matrices, automatically, using patsy.dmatrices function.

 - patsy.dmatrices takes a formula (in string form) as a first argument, and a dictionary-like object with data arrays for the response variables as second arguments.

In [14]:
import patsy
import numpy as np

y = np.array([1, 2, 3, 4, 5])
x1 = np.array([6, 7, 8, 9, 10])
x2 = np.array([11, 12, 13, 14, 15])
data = {'y':y, 'x1':x1, 'x2':x2}

y, X = patsy.dmatrices('y ~ 1 + x1 + x2 + x1*x2', data)

print(y)
print(X)

[[1.]
 [2.]
 [3.]
 [4.]
 [5.]]
[[  1.   6.  11.  66.]
 [  1.   7.  12.  84.]
 [  1.   8.  13. 104.]
 [  1.   9.  14. 126.]
 [  1.  10.  15. 150.]]


**Understanding patsy Formulae**

Let's understand few patsy formulae provided below.

 - 'y ~ x' : y is linearly dependent on x. ~ symbol separates dependent variable from independent variable terms. It is also equivalent to 'y ~ 1 + x'.

 - 'y ~ x1 + x2' : y is a linear combination of x1 and x2. + sign is used to denote the union of terms.

 - y ~ x1*x2 : x1*x2 is an interaction term that includes all lower order terms. Hence formula is equivalent to y ~ 1 + x1 + x2 + x1*x2.

  - 'y ~ np.log(x1)': Often numpy functions can be used to transform terms in the expression.

 - 'y ~ I(x1 + x2)': I is the identify function, used to escape arithmetic expressions and are evaluated.

 - 'y ~ C(x1)': Treats the variable x1 as a categorical variable.

**Example Datasets**
 - statsmodels contain few popular example datasets, which can be used to explore various utilities of the package.

 - The example data sets are available in datasets module.

 - Each dataset is associated with special variables like SOURCE, DESCSHORT, DESCLONG, that provide more info about the dataset.

 - A dataset can be loaded using load function, and its data can be accessed using data attribute in the form of Numpy's recarray.

**Loading Example Datasets**
 - The below example shows loading of popular breast cancer dataset and accessing its data as numpy array.

In [15]:
import statsmodels.api as sm

bc_cancer_set = sm.datasets.cancer

bc_cancer = bc_cancer_set.load()

bc_cancer_data = bc_cancer.data

print(type(bc_cancer_data))

<class 'numpy.recarray'>


**Loading R Datasets**

statsmodels package provides access to many example datasets of R, listed at R Datasets repository (https://vincentarelbundock.github.io/Rdatasets/datasets.html)

The below example loads Icecream dataset from Ecdat package.

The data is accessible via data attribute, as a pandas data frame.

In [16]:
import statsmodels.api as sm

icecream_data = sm.datasets.get_rdataset('Icecream', 'Ecdat')
data = icecream_data.data

print(icecream_data.data.shape)

(30, 4)


# StatsModel

Let's understand how to fit a linear regression model for Icecream dataset available from R Data Repository.

In [17]:
import statsmodels.api as sm

icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")

icecream_data = icecream.data

print(icecream_data.columns)

Index(['cons', 'income', 'price', 'temp'], dtype='object')


The icecream_data dataset is a pandas data frame. It contains four variables: cons (consumption), income, price, and temp (temperature).

**Choosing a Model**

Initially, let's model consumption with price and temperature, as a linear model.

The patsy formula for assumed model is: cons ~ price + temp

**Creating, Fitting and viewing Summary of a Model**

In [18]:
import statsmodels.formula.api as smf

linear_model1 = smf.ols('cons ~ price + temp', icecream_data)
linear_result1 = linear_model1.fit()
print(linear_result1.summary())

                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.633
Model:                            OLS   Adj. R-squared:                  0.606
Method:                 Least Squares   F-statistic:                     23.27
Date:                Wed, 21 Sep 2022   Prob (F-statistic):           1.34e-06
Time:                        09:59:05   Log-Likelihood:                 54.607
No. Observations:                  30   AIC:                            -103.2
Df Residuals:                      27   BIC:                            -99.01
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      0.5966      0.258      2.309      0.0

**Analyzing the Model**

The R-squared value of 0.633 suggests that model is not a proper fit.

The probability value of coefficient price is high, i.e., 0.141. This accepts the null-hypothesis: the value of price coefficient is equal to zero.

Hence, the variable price does not affect cons variable.

**Model Recreation**

Now let's create a new model by considered income and temp dependent variables.


In [19]:
linear_model2 = smf.ols('cons ~ income + temp', icecream_data)

linear_result2 = linear_model2.fit()

print(linear_result2.summary())

                            OLS Regression Results                            
Dep. Variable:                   cons   R-squared:                       0.702
Model:                            OLS   Adj. R-squared:                  0.680
Method:                 Least Squares   F-statistic:                     31.81
Date:                Wed, 21 Sep 2022   Prob (F-statistic):           7.96e-08
Time:                        09:59:05   Log-Likelihood:                 57.742
No. Observations:                  30   AIC:                            -109.5
Df Residuals:                      27   BIC:                            -105.3
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.1132      0.108     -1.045      0.3

R-squared value improved to 0.702, suggesting this is a better model than previous one.

Probability values of income and temp are low, indicating they are highly significant to cons.

The probability value of Intercept is too high.

Now let's create a new model without considering intercept term.

In [20]:
linear_model3 = smf.ols('cons ~ -1 + income + temp', icecream_data)

linear_result3 = linear_model3.fit()

print(linear_result3.summary())

                                 OLS Regression Results                                
Dep. Variable:                   cons   R-squared (uncentered):                   0.990
Model:                            OLS   Adj. R-squared (uncentered):              0.990
Method:                 Least Squares   F-statistic:                              1426.
Date:                Wed, 21 Sep 2022   Prob (F-statistic):                    6.77e-29
Time:                        09:59:05   Log-Likelihood:                          57.146
No. Observations:                  30   AIC:                                     -110.3
Df Residuals:                      28   BIC:                                     -107.5
Df Model:                           2                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [21]:
# Load the R dataset mtcars.
# Capture the data as pandas dataframes.
# Build a linear regression model with independent variable wt, and dependent variable mpg
# Fit the model with data, and dispaly the R-Squared value

import statsmodels.api as sm
import statsmodels.formula.api as smf

cars = sm.datasets.get_rdataset("mtcars", "datasets")
data = cars.data

linear_model1 = smf.ols('mpg ~ wt', data)
linear_result1 = linear_model1.fit()

print(round(linear_result1.rsquared,3))


0.753


To improve the rsquared values, taking log of both dependent and independent variable

In [22]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np

cars = sm.datasets.get_rdataset("mtcars", "datasets")
data = cars.data

linear_model1 = smf.ols('np.log(mpg) ~ np.log(wt)', data)
linear_result1 = linear_model1.fit()

print(round(linear_result1.rsquared,3))

0.806


**Discrete Regression - Introduction**

A discrete dependent variable takes few possible outcome values and is not normally distributed.

Hence, linear regression cannot be applied to a discrete variable.

**Discrete Regression models**

statsmodels provide the following classes to work with discrete regression problems.

Logit: for Logistic Regression
MNLogit: for Multinomial Logistic Regression
Poisson: for Poisson Regression


Now, let's understand how to perform logistic regression using statsmodels with the following steps:

 - Download the popular iris data set (containing data of 3 species) from R repository.
 - Subset the data of only two species.
 - Perform transformations, if required.
 - Define a patsy formula and create a model using logit.
 - Fit the model with supplied data.
 - View summary of the model.


 **Loading the dataset**

In [23]:
import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd

iris = sm.datasets.get_rdataset("iris").data 
iris.info()
iris.Species.unique() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Sepal.Length  150 non-null    float64
 1   Sepal.Width   150 non-null    float64
 2   Petal.Length  150 non-null    float64
 3   Petal.Width   150 non-null    float64
 4   Species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


array(['setosa', 'versicolor', 'virginica'], dtype=object)

As seen in the previous example, iris dataset contains details of three types of species.

In a logistic regression, the response variable refers to only two type of categories. Hence, data of two species is filtered and considered in further steps.

In [24]:
# Sub setting
iris_subset = iris[(iris.Species == "versicolor") | (iris.Species == "virginica")].copy()

print(iris_subset.Species.unique())

['versicolor' 'virginica']


**Applying Transformations**

A binary variable corresponding to two species is created using map function as shown below.

In [25]:
df_subset = iris[(iris.Species == "versicolor") | (iris.Species == "virginica" )].copy()
# period characters, present in column names are replaced with underscore characters.

df_subset.Species = df_subset.Species.map({"versicolor": 1, "virginica": 0}) 

df_subset.rename(columns={"Sepal.Length": "Sepal_Length", "Sepal.Width": "Sepal_Width",	
                          "Petal.Length": "Petal_Length", "Petal.Width": "Petal_Width"}, inplace=True) 

A logistic regression model, explaining relationship of Species variable with Petal_Length and Petal_width is generated using logit function.

The appropriate patsy formula is passed as argument to logit function.

In [26]:
model = smf.logit("Species ~ Petal_Length + Petal_Width", data=df_subset)
result = model.fit()
print(result.summary())

Optimization terminated successfully.
         Current function value: 0.102818
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                Species   No. Observations:                  100
Model:                          Logit   Df Residuals:                       97
Method:                           MLE   Df Model:                            2
Date:                Wed, 21 Sep 2022   Pseudo R-squ.:                  0.8517
Time:                        09:59:06   Log-Likelihood:                -10.282
converged:                       True   LL-Null:                       -69.315
Covariance Type:            nonrobust   LLR p-value:                 2.303e-26
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       45.2723     13.612      3.326      0.001      18.594      71.951
Petal_Length    -5.754

**Analyzing the Model**

The model summary indicates that both Petal_Length and Petal_Width are statistically significant.

For every unit change in Petal_Length, the log odds of being versicolor decreases by 5.75.

For every unit change in Petal_Width, the log odds of being versicolor decreases by 10.44.

Once satisfied with the model, it can be used to predict response variable value for new values of independent variables.

**Predicting Response Variable**

Now, let's create 20 observations having random values of Petal_Length and Petal_Width.

Use these 20 observations to predict values for Species.

In [27]:
df_new = pd.DataFrame({"Petal_Length": np.random.randn(20)*0.5 + 5,
                       "Petal_Width": np.random.randn(20)*0.5 + 1.7})
df_new["P-Species"] = result.predict(df_new)
df_new["P-Species"].head(3)

df_new["Species"] = (df_new["P-Species"] > 0.5).astype(int)
df_new.head()

Petal_Length  Petal_Width  P-Species  Species
0      4.799561     2.508475   0.000193        0
1      5.412003     1.951370   0.001914        0
2      4.718847     2.479403   0.000416        0
3      5.977439     1.754701   0.000578        0
4      4.334024     1.090128   0.999869        1

In [28]:
# Load r dataset biopsy from MASS package
# Capture the data as pandas dataframe
# Rename the column class to Class
# Transform the Class coloumn values benign to 0 and malignant to 1.
# Build a logistic regression model with independent variable V1 and dependent variable Class
# Fit the model with data, and display the pseudo r squared value

import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd

df = sm.datasets.get_rdataset("biopsy", "MASS").data
df.rename(columns = {'class':'Class'}, inplace = True)

df.Class = df.Class.map({"benign":0, "malignant":1})

model = smf.logit("Class ~ V1", data=df)
result = model.fit() 
print(round(result.prsquared,4))

Optimization terminated successfully.
         Current function value: 0.331941
         Iterations 7
0.4847


**Poisson Model**

Poisson Model describes a process where dependent variable refers to success count of many attempts and each attempt has a very low probability of success.

Let's understand how to fit a Poisson regression model for a data set available at UCLA repository.

The dataset contains details of a number of awards earned, type of program enrolled, and score obtained in final math exam by students at a high school.

**Loading the Dataset**
The dataset is fetched as a pandas data frame as shown belo

In [29]:
import pandas as pd

awards_df = pd.read_csv("https://stats.idre.ucla.edu/stat/data/poisson_sim.csv")

print(awards_df.head(3))

# There are three type of programs enrolled by a student : 1 - "General", 2 - "Academic", 3 - "Vocational"

    id  num_awards  prog  math
0   45           0     3    41
1  108           0     1    41
2   15           0     3    44


In [30]:
import statsmodels.formula.api as smf

poisson_model = smf.poisson('num_awards ~ math + C(prog)', awards_df)
poisson_model_result = poisson_model.fit()
print(poisson_model_result.summary())

Optimization terminated successfully.
         Current function value: 0.913761
         Iterations 6
                          Poisson Regression Results                          
Dep. Variable:             num_awards   No. Observations:                  200
Model:                        Poisson   Df Residuals:                      196
Method:                           MLE   Df Model:                            3
Date:                Wed, 21 Sep 2022   Pseudo R-squ.:                  0.2118
Time:                        09:59:07   Log-Likelihood:                -182.75
converged:                       True   LL-Null:                       -231.86
Covariance Type:            nonrobust   LLR p-value:                 3.747e-21
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -5.2471      0.658     -7.969      0.000      -6.538      -3.957
C(prog)[T.2]     1.0839

**Analyzing Model Summary**


The coefficient for math variable is 0.07, which means for every one unit increase in math, the log count increases by 0.07.

Having enrolled for prog=2, i.e., "Academic", instead of "Generic" program, changes the log count by 1.08.

Having enrolled for prog=3, i.e., "Vocational", instead of "Generic" program, changes the log count by 0.37.

In [31]:
# Load r dataset Insurance from MASS package
# Capture the data as pandas dataframe
# Build a poisson regression model with log of an independent variable Holders and dependent variable Claims
# Fit the model with data, and display the sum of residuals

import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd

df = sm.datasets.get_rdataset("Insurance", "MASS").data
df = {'holders':df['Holders'],'claims':df['Claims']}
model = smf.poisson("claims ~ np.log(holders)", df)
result = model.fit() 
print(np.cumsum(result.resid))


Optimization terminated successfully.
         Current function value: 3.468160
         Iterations 7
0     5.224177e+00
1    -1.766208e+00
2    -2.132062e+01
3    -6.637946e+01
4    -4.804676e+01
          ...     
59   -1.579198e+01
60   -1.674158e+01
61   -1.465717e+01
62   -1.236967e+01
63    6.679102e-13
Length: 64, dtype: float64


**ANOVA Example**

Let's now understand how to perform ANOVA of a fitted regression model.

For this example, let's once again consider popular Icecream dataset, used in linear regression section.

In [32]:
# loading the model
import statsmodels.api as sm

icecream = sm.datasets.get_rdataset("Icecream", "Ecdat")
icecream_data = icecream.data

# building the model

import statsmodels.formula.api as smf

model1 = smf.ols('cons ~ temp', icecream_data).fit()

#A linear regression model of icecream consumption with temperature is built using ols method.

# Now let's frame a null hypothesis: Value of B1, i.e., the coefficient of temp variable is zero.

**Creating ANOVA Table**

anova_lm method from anova module of statsmodels library can be used for performing ANOVA.

In [33]:
from statsmodels.stats import anova
print(anova.anova_lm(model1))

            df    sum_sq   mean_sq         F        PR(>F)
temp       1.0  0.075514  0.075514  42.27997  4.789215e-07
Residual  28.0  0.050009  0.001786       NaN           NaN


The obtained F-statistic is 42.27 and has a very low probability.

Hence, the null hypothesis can be rejected. i.e., B1 is not equal to zero.

**Building New Model**

Now let's create a new model with two independent variables and one response variable.

With more that one independent variable, the null hypothesis is stated as Coefficients of all independent variables are zero. i.e B1 = 0, and B2 = 0.

The alternative hypothesis will be that atleast one of the parameters Bj != 0 where j takes the values 1, 2, ...

In [34]:
model2 = smf.ols('cons ~ income + temp', icecream_data).fit()
print(anova.anova_lm(model2))

            df    sum_sq   mean_sq          F        PR(>F)
income     1.0  0.000288  0.000288   0.208231  6.518069e-01
temp       1.0  0.087836  0.087836  63.413711  1.470071e-08
Residual  27.0  0.037399  0.001385        NaN           NaN


The above table shows F-statistic corresponding to each independent variable.

Calculation of F-statistic for the model is shown as following

**Determining F-Value**

F-statistic is defined as Mean square of model/Mean square of residuals.

Mean square of model is sum of squares values of all variables/degrees of freedom i.e (0.000288 + 0.087836)/2 = 0.044062.

Hence, the F-statistic of model is (0.044062 / 0.001385) = 31.813.

The probability of obtained F-statistic is 7.96*e-08 and is computed as shown below:

In [35]:
from scipy import stats

print(stats.f.sf(31.81, 2, 27))

7.959504548627583e-08


Since the p-value is low, null hypothesis is rejected.

**Comparing Two Models**

ANOVA can also be used to compare two nested or related models.

model1 regresses consumption with temperature and model2 with temperature and income.

Below code verifies if the decrease in residuals sum of squares is significant or not

In [36]:
print(anova.anova_lm(model1, model2))

   df_resid       ssr  df_diff   ss_diff         F    Pr(>F)
0      28.0  0.050009      0.0       NaN       NaN       NaN
1      27.0  0.037399      1.0  0.012611  9.104375  0.005506


p-value i.e 0.005 from above table suggests the decrease in residual sum of squares in model2, compared to model1 is significant.

In [37]:
# load the rdatasets mtcars
# capture thre data as pandas dataframe
# build a linear regresion model with independent variable wt and dependent variable mpg
# fit the model with data
# perform ANOVA on the linear model obtained in the previous step
# display f-statistics value

import statsmodels.api as sm
import statsmodels.formula.api as smf
import numpy as np
import pandas as pd
from statsmodels.stats import anova

df = sm.datasets.get_rdataset("mtcars", "datasets").data
model = smf.ols("mpg ~ wt", df)
result = model.fit() 
print(round(anova.anova_lm(result).F[0],6))

# using log on both dependednt and independent variable to  increase f-statistics
model = smf.ols("np.log(mpg) ~ np.log(wt)", df)
result = model.fit() 
print(round(anova.anova_lm(result).F[0],6))

91.375325
124.359619
